### PR TITLE
Remote Bazefetcher source

### PR DESCRIPTION
**Pull request by JensGM**

### CODE REVIEW

File: camille/source/bazefetcher.py [hunk 1]
```diff
@@ -1,17 +1,209 @@
-import os
-from os.path import join, isdir, basename
-import re
+from ..util import utcdate
+from abc import ABC, abstractmethod
```
Comment (mrSundvoll) [inline camille/source/bazefetcher.py:2]: Split on two lines?

File: camille/source/bazefetcher.py [hunk 2]
```diff
@@ -51,47 +243,29 @@ def _tidy_frame(df, tzinfo):
     df.sort_index(inplace=True)
 
 
-def _get_files(src_dirs, tag, fn_regex, date_pred):
+def _get_files(io, fn_regex, date_pred):
     """
-    Gets all possible files with tag under src_dirs which satisfy
+    Gets all possible files with tag under io which satisfy
     fn_regex and date_func
```
Comment (mrSundvoll) [inline camille/source/bazefetcher.py:249]:
```suggestion
    fn_regex and date_pred
```

**Review by mrSundvoll — COMMENTED** [review]
I don't comprehend everything that is going on, but here are some comments and suggestions at least

File: camille/source/bazefetcher.py [hunk 3]
```diff
@@ -1,17 +1,209 @@
...
+
+        def iterdir(self):
+            return (self / f for f in self.owner.sftp.listdir(str(self.path)))
+
+        def open(self,
+                 mode='r',
+                 buffering=-1,
+                 encoding=None,
+                 errors=None,
+                 newline=None):
```
[the remark this answers is not in the document]
Comment (JensGM) [reply camille/source/bazefetcher.py:77]: It matches pathlib.Path.open

File: camille/source/bazefetcher.py [hunk 4]
```diff
@@ -1,17 +1,209 @@
...
+            return f'{self.__class__.__name__}<{str(self.path)}>'
+
+    @classmethod
+    def parse_path(cls, path):
+        xs = path
+        username = None
+        if '@' in path:
+            username, xs = xs.split('@')
+        host, p = xs.split(':')
+        return pathlib.Path(p), host, username
```
[the remark this answers is not in the document]
Comment (JensGM) [reply camille/source/bazefetcher.py:125]: Nope!

File: camille/source/bazefetcher.py [hunk 5]
```diff
@@ -1,17 +1,209 @@
...
+                          port=self.port,
+                          username=self.username,
+                          allow_agent=False,
+                          look_for_keys=True)
+        return conection
+
+    def __enter__(self):
+        with self._lock:
+            if self._connection is None:
+                self._connection = self._crate_connection()
```
Comment (JensGM) [inline camille/source/bazefetcher.py:163]:
```suggestion
                self._connection = self._create_connection()
```

File: tests/source/test_bazefetcher.py [hunk 6]
```diff
@@ -13,6 +17,26 @@
...
+
+@contextlib.contextmanager
+def mock_remote_ssh():
+    with contextlib.ExitStack() as exit_stack:
+        users = {'ssh-user': 'tests/test_data/ssh/id_rsa'}
+        server = exit_stack.enter_context(mockssh.Server(users))
+        client = exit_stack.enter_context(server.client('ssh-user'))
+
+        connect_mock = exit_stack.enter_context(mock.patch(
+            'camille.source.bazefetcher.RemoteIO._crate_connection'
```
Comment (JensGM) [inline tests/source/test_bazefetcher.py:29]:
```suggestion
            'camille.source.bazefetcher.RemoteIO._create_connection'
```

**Pull request merged**